src/platform/graphics/dynamic-buffers.js

```diff
@@ -0,0 +1,218 @@
+import { Debug } from '../../core/debug.js';
+import { math } from '../../core/math/math.js';
+
+/**
+ * A base class representing a single per platform buffer.
+ *
+ * @ignore
+ */
+class DynamicBuffer {
+    /** @type {import('./graphics-device.js').GraphicsDevice} */
+    device;
+
+    constructor(device) {
+        this.device = device;
+    }
+}
+
+/**
+ * A container for storing the used areas of a pair of staging and gpu buffers.
+ *
+ * @ignore
+ */
+class UsedBuffer {
+    /** @type {DynamicBuffer} */
+    gpuBuffer;
+
+    /** @type {DynamicBuffer} */
+    stagingBuffer;
+
+    /**
+     * The beginning position of the used area that needs to be copied from staging to to the GPU
+     * buffer.
+     *
+     * @type {number}
+     */
+    offset;
+
+    /**
+     * Used byte size of the buffer, from the offset.
+     *
+     * @type {number}
+     */
+    size;
+}
+
+/**
+ * A container for storing the return values of an allocation function.
+ *
+ * @ignore
+ */
+class DynamicBufferAllocation {
+    /**
+     * The storage access to the allocated data in the staging buffer.
+     *
+     * @type {Int32Array}
+     */
+    storage;
+
+    /**
+     * The gpu buffer this allocation will be copied to.
+     *
+     * @type {DynamicBuffer}
+     */
+    gpuBuffer;
+
+    /**
+     * Offset in the gpuBuffer where the data will be copied to.
+     *
+     * @type {number}
+     */
+    offset;
+}
+
+/**
+ * The DynamicBuffers class provides a dynamic memory allocation system for uniform buffer data,
+ * particularly for non-persistent uniform buffers. This class utilizes a bump allocator to
+ * efficiently allocate aligned memory space from a set of large buffers managed internally. To
+ * utilize this system, the user writes data to CPU-accessible staging buffers. When submitting
+ * command buffers that require these buffers, the system automatically uploads the data to the GPU
+ * buffers. This approach ensures efficient memory management and smooth data transfer between the
+ * CPU and GPU.
+ *
+ * @ignore
+ */
+class DynamicBuffers {
+    /**
+     * Allocation size of the underlying buffers.
+     *
+     * @type {number}
+     */
+    bufferSize;
+
+    /**
+     * Internally allocated gpu buffers.
+     *
+     * @type {DynamicBuffer[]}
+     */
+    gpuBuffers = [];
+
+    /**
+     * Internally allocated staging buffers (CPU writable)
+     *
+     * @type {DynamicBuffer[]}
+     */
+    stagingBuffers = [];
+
+    /**
+     * @type {UsedBuffer[]}
+     */
+    usedBuffers = [];
+
+    /**
+     * @type {UsedBuffer}
+     */
+    activeBuffer = null;
+
+    /**
+     * Create the system of dynamic buffers.
+     *
+     * @param {import('./graphics-device.js').GraphicsDevice} device - The graphics device.
+     * @param {*} bufferSize - The size of the underlying large buffers.
+     * @param {*} bufferAlignment - Alignment of each allocation.
```

Comment:
Aren't the types here `number`?

Reply:
yep, I missed filling those in, done now, thanks